util/stopper.go

```diff
@@ -106,6 +106,9 @@ func (s *Stopper) StartTask() bool {
 		// guarantee that that's still the case now.
 		s.mu.Lock()
 		defer s.mu.Unlock()
+		if atomic.LoadInt32(&s.draining) == 1 {
+			return false
+		}
```

Comment:
I don't know why you're checking for draining again. It could always be draining after you check. Also the comments about the deadlocking above are scary because you could start draining just before you attempt to hold on to the lock.